# Data is from EBay. Classified section of the German eBay Website

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding= 'Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Names are separated by underscores
### Price has dollar sign
### odometer has 'km'
### HAH it says 'nein'. German


In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
        'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
        'odometer', 'registration_month', 'fuel_type', 'brand',
        'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
        'last_seen']
autos.columns = cols

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Changed the format of the header from camelcase to snakecase

In [9]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [10]:
autos['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

In [11]:
autos['registration_month'].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

### Price and Odometer can probably lose the dollar sign and km respectively 

In [12]:
autos['price'] = (autos['price']
                  .str.replace('$','')
                  .str.replace(',','')
                  .astype(float)
                 )

In [13]:
autos['odometer'] = (autos['odometer']
                     .str.replace('km','')
                     .str.replace(',','')
                     .astype(float)
                    )

In [14]:
autos.rename(columns = {'odometer': 'odometer_km'}, inplace=True)

## Explore the data, starting with Odometer_km and price

In [15]:
autos['odometer_km'].value_counts(ascending = True)

10000.0       264
20000.0       784
30000.0       789
40000.0       819
5000.0        967
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64

In [16]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [17]:
autos['price'].value_counts().sort_index(ascending = False)

99999999.0       1
27322222.0       1
12345678.0       3
11111111.0       2
10000000.0       1
3890000.0        1
1300000.0        1
1234566.0        1
999999.0         2
999990.0         1
350000.0         1
345000.0         1
299000.0         1
295000.0         1
265000.0         1
259000.0         1
250000.0         1
220000.0         1
198000.0         1
197000.0         1
194000.0         1
190000.0         1
180000.0         1
175000.0         1
169999.0         1
169000.0         1
163991.0         1
163500.0         1
155000.0         1
151990.0         1
              ... 
66.0             1
65.0             5
60.0             9
59.0             1
55.0             2
50.0            49
49.0             4
47.0             1
45.0             4
40.0             6
35.0             1
30.0             7
29.0             1
25.0             5
20.0             4
18.0             1
17.0             3
15.0             2
14.0             1
13.0             2
12.0             3
11.0        

In [18]:
autos.loc[autos['price'] == 0].head()

date_crawled                                               name  \
27  2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71  2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80  2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87  2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99  2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   

    seller offer_type  price   abtest vehicle_type  registration_year  \
27  privat    Angebot    0.0  control          NaN               2005   
71  privat    Angebot    0.0  control          NaN               1990   
80  privat    Angebot    0.0  control        coupe               1999   
87  privat    Angebot    0.0  control          NaN               2000   
99  privat    Angebot    0.0  control       cabrio               2011   

    gearbox  power_ps    model  odometer_km  registration_month fuel_type  \
27      NaN         0      NaN     150000.0                   0       NaN   
71  manuell         0      NaN       5000.0                   0    benzin   
80  manuell        99  primera     150000.0                   3    benzin   
87      NaN         0      5er     150000.0                   0       NaN   
99  manuell         0  2_reihe      60000.0                   7    diesel   

      brand unrepaired_damage           ad_created  nr_of_pictures  \
27     ford               NaN  2016-03-27 00:00:00               0   
71     opel               NaN  2016-03-28 00:00:00               0   
80   nissan                ja  2016-03-09 00:00:00               0   
87      bmw               NaN  2016-03-29 00:00:00               0   
99  peugeot              nein  2016-04-05 00:00:00               0   

    postal_code            last_seen  
27        66701  2016-03-27 18:45:01  
71         4552  2016-04-07 01:45:48  
80        66903  2016-03-09 16:43:50  
87        82256  2016-04-06 21:18:15  
99        99735  2016-04-07 12:17:34

In [19]:
# Weird numbers due to weird distribtuion of prices in the price columns
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [20]:
autos['price'].max()

99999999.0

### Getting rid of the rows that seem like outliers

In [21]:
autos[autos['price'].between(350001,100000000000)].head()

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   

       seller offer_type       price   abtest vehicle_type  registration_year  \
514    privat    Angebot    999999.0     test        kombi               2009   
2897   privat    Angebot  11111111.0     test    limousine               1973   
7814   privat    Angebot   1300000.0  control        coupe               1992   
11137  privat    Angebot  10000000.0  control        coupe               1960   
22947  privat    Angebot   1234566.0  control        kombi               1999   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
514      manuell       101   focus     125000.0                   4    benzin   
2897     manuell        48  escort      50000.0                   3    benzin   
7814         NaN         0     NaN      50000.0                  12       NaN   
11137    manuell       368     NaN     100000.0                   1    benzin   
22947  automatik       190     NaN     150000.0                   2    diesel   

                brand unrepaired_damage           ad_created  nr_of_pictures  \
514              ford              nein  2016-03-17 00:00:00               0   
2897             ford              nein  2016-03-12 00:00:00               0   
7814   sonstige_autos              nein  2016-04-04 00:00:00               0   
11137  sonstige_autos              nein  2016-03-29 00:00:00               0   
22947             bmw               NaN  2016-03-22 00:00:00               0   

       postal_code            last_seen  
514          12205  2016-04-06 07:17:35  
2897         94469  2016-03-12 22:45:27  
7814         60598  2016-04-05 11:34:11  
11137        73033  2016-04-06 21:18:11  
22947        17454  2016-04-02 03:17:32

In [22]:
bins = np.arange(-100, 10000, 100)
#autos['price'].sort_index().value_counts(bins=bins)
s = autos.groupby(pd.cut(autos['price'], bins=bins)).size()
print(s)

price
(-100, 0]       1421
(0, 100]         475
(100, 200]       711
(200, 300]       895
(300, 400]       898
(400, 500]      1270
(500, 600]      1177
(600, 700]      1115
(700, 800]      1200
(800, 900]      1136
(900, 1000]     1701
(1000, 1100]     539
(1100, 1200]    1089
(1200, 1300]     920
(1300, 1400]     771
(1400, 1500]    1275
(1500, 1600]     645
(1600, 1700]     669
(1700, 1800]     728
(1800, 1900]     572
(1900, 2000]    1147
(2000, 2100]     264
(2100, 2200]     568
(2200, 2300]     554
(2300, 2400]     423
(2400, 2500]     995
(2500, 2600]     386
(2600, 2700]     416
(2700, 2800]     562
(2800, 2900]     455
                ... 
(6900, 7000]     433
(7000, 7100]      27
(7100, 7200]     136
(7200, 7300]     149
(7300, 7400]     122
(7400, 7500]     356
(7500, 7600]      96
(7600, 7700]     105
(7700, 7800]     178
(7800, 7900]     210
(7900, 8000]     424
(8000, 8100]      29
(8100, 8200]      95
(8200, 8300]     106
(8300, 8400]      79
(8400, 8500]     332
(8500, 

In [23]:
autos.drop(autos[autos['price'].between(350001,100000000000)].index, inplace = True)

In [24]:
autos.shape

(49986, 20)

### Dropped all records with a price greater than 350K. Seemed like alot of those prices were inflated or default values (values like 1234566)

## Dates and Stuff

In [25]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49986 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          49986 non-null object
name                  49986 non-null object
seller                49986 non-null object
offer_type            49986 non-null object
price                 49986 non-null float64
abtest                49986 non-null object
vehicle_type          44894 non-null object
registration_year     49986 non-null int64
gearbox               47310 non-null object
power_ps              49986 non-null int64
model                 47233 non-null object
odometer_km           49986 non-null float64
registration_month    49986 non-null int64
fuel_type             45509 non-null object
brand                 49986 non-null object
unrepaired_damage     40163 non-null object
ad_created            49986 non-null object
nr_of_pictures        49986 non-null int64
postal_code           49986 non-null int64
last_seen             49986 non-null object
dtypes: float6

In [26]:
autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [27]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

### Date crawled falls after April 4th. Generally, % of volume per date is between 3% to 4%. Occasional % of volume drops below 2%.

In [28]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = True)

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033229
2016-03-10    0.031869
2016-03-11    0.032789
2016-03-12    0.036610
2016-03-13    0.016925
2016-03-14    0.035230
2016-03-15    0.033749
2016-03-16    0.030008
2016-03-17    0.031189
2016-03-18    0.013724
2016-03-19    0.033849
2016-03-20    0.037871
2016-03-21 

## Frequency of Ad Creation appears to increase in or near March 2016

In [29]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = True)

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64

## Almost 50% of ads were last seen on April 5, 6, and 7. The rest of the ads have a more equal distribution across all dates 

In [30]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

## Pretty weird to have a minimum of 1000 and max of 9999 for a year

In [31]:
autos[autos['registration_year']>2016]['ad_created'].head()

10    2016-03-14 00:00:00
55    2016-03-06 00:00:00
65    2016-04-04 00:00:00
68    2016-04-03 00:00:00
84    2016-03-27 00:00:00
Name: ad_created, dtype: object

In [32]:
autos[autos['registration_year']<1900].shape

(6, 20)

In [33]:
autos = autos[autos['registration_year'].between(1900,2016)]

## Including only car ads where the cars registration year is between 1900 and 2016

In [34]:
autos['registration_year'].value_counts(normalize=True)

2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
2006    0.056377
2001    0.056273
2002    0.052753
1998    0.051087
2007    0.047984
2008    0.046464
2009    0.043673
1997    0.042236
2011    0.034030
2010    0.033260
1996    0.030073
2012    0.027553
2016    0.027408
1995    0.027324
2013    0.016786
2014    0.013850
1994    0.013745
1993    0.009268
2015    0.008310
1990    0.008226
1992    0.008122
1991    0.007414
1989    0.003770
1988    0.002957
1985    0.002166
          ...   
1966    0.000458
1977    0.000458
1975    0.000396
1969    0.000396
1965    0.000354
1964    0.000250
1910    0.000187
1963    0.000187
1959    0.000146
1961    0.000125
1956    0.000104
1958    0.000083
1937    0.000083
1962    0.000083
1950    0.000062
1954    0.000042
1941    0.000042
1934    0.000042
1957    0.000042
1951    0.000042
1955    0.000042
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.0000

## Working on Aggregating by Brand

In [35]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
smart             0.013912
citroen           0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
dacia             0.002562
daihatsu          0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

In [36]:
s = autos['brand'].value_counts(normalize=True)
s[s>0.01].index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'sonstige_autos'],
      dtype='object')

In [37]:
brand_types = s[s>0.01].index

In [38]:
print(brand_types)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'sonstige_autos'],
      dtype='object')


In [50]:
autos.loc[autos['brand'] == b]

date_crawled                                               name  \
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
40     2016-03-07 14:50:03  VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...   
41     2016-03-10 10:46:08  Passat_3b_1.9_TDI_Highline__angemeldet_mit_tuv...   
42     2016-03-22 21:37:03  Vw_Polo_l.0_/60_PS_Blue_Motion_Technologie_Son...   
57     2016-03-12 11:46:56    Volkswagen_Passat_Variant_2.0_TDI_DPF_Trendline   
58     2016-03-30 12:25:18                                Volkswagen_Golf_1.4   
60     2016-03-23 21:55:29                              VW_Vento_1_8_Tuev_NEU   
62     2016-03-12 17:41:11                  Volkswagen_Golf_Plus_1.6_FSI_Goal   
73     2016-03-18 23:41:21                        VW_T3_Doka_1_7D_TÜV_07/2017   
96     2016-03-07 20:56:44              Volkswagen_Golf_Variant_1.9_TDI_Joker   
104    2016-03-12 12:57:24                Volkswagen_Multivan_T4_TDI_Highline   
112    2016-03-24 00:32:33                           Volkswagen_Polo_75_Servo   
114    2016-03-23 17:51:11  Volkswagen_Passat_Variant_2.0_TDI_BlueMotion_T...   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
120    2016-03-24 22:56:04  Golf_VII_1_4_TSI_Highline_Autom_Panorama_Alcan...   
124    2016-03-16 14:51:35                     Passat_2.0_TDI_DPF_Comfortline   
127    2016-03-21 15:49:37                    Volkswagen_polo_6n_tuev_11/2016   
128    2016-03-10 12:06:20                         Volkswagen_Lupo_1.0_Oxford   
147    2016-03-15 14:52:55                          VW_Polo_Ersatzteilspender   
148    2016-03-17 07:57:07        Polo_6N_mit_tuev_bis_03._2017_fuer_500_euro   
151    2016-03-27 09:55:44              Volkswagen_Golf_Cabrio_2.0_Generation   
154    2016-04-01 13:56:00            Volkswagen_Transporter_T4_TDI_LKW/Wohmo   
155    2016-03-07 18:45:47                                  VW_Touran__140_PS   
161    2016-03-30 21:54:12                  Volkswagen_Touareg_3.0_V6_TDI_DPF   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
...                    ...                                                ...   
49845  2016-03-21 00:50:15  Schlachte_VW_Sharan_vr6_Automatik___no_GTI_16V...   
49847  2016-04-03 22:58:18  Volkswagen_Golf_1.4_TSI_BlueMotion_Technology_...   
49850  2016-03-21 06:36:54                                   Vw_Passat_2_0tdi   
49852  2016-04-01 04:02:25                                   TOP__Golf_3_1.8l   
49857  2016-03-21 17:40:01  Volkswagen_Golf_1.6_TDI_BlueMotion_Technology_...   
49859  2016-03-26 01:00:17           Golf_4___5Tuerig_schwarz_Spezial_Edition   
49864  2016-03-08 20:58:47                     Golf_5_1_9_tdi.....2jahre_tuev   
49866  2016-03-29 23:43:12                         Golf_III_zum_Ausschlachten   
49867  2016-03-12 17:59:44    VOLKSWAGEN_Tiguan_1.4_TSI_Sport_&_Style_4Motion   
49869  2016-03-17 19:56:31                   Volkswagen_Sharan_1.9_TDI_Family   
49873  2016-04-02 09:55:27                        VW_POLO_6_N_mit_TÜV_12.2016   
49875  2016-03-29 13:55:06               Golf_5_Variant_ab_April_zu_verkaufen   
49881  2016-04-02 21:52:37                          Volkswagen_Golf_Syncro_CL   
49892  2016-03-14 14:57:03                                         Vw_T_4_Bus   
49896  2016-03-25 19:51:40                                   Vw_golf_1_cabrio   
49905  2016-03-23 20:54:05                                Volkswagen_Lupo_1.4   
49925  2016-03-30 11:46:18  Volkswagen_Passat_Variant_1.9_TDI_Comfortl._mi...   
49926  2016-03-13 00:36:56                   

In [54]:
brand_types = s[s>0.01].index
avg_price = {}

for b in brand_types:
    brand_only = autos[autos['brand'] == b]
    avg_price[b] = brand_only['price'].mean()
    
avg_price

{'audi': 9093.65003615329,
 'bmw': 8102.536248343744,
 'citroen': 3699.935628742515,
 'fiat': 2711.8011272141707,
 'ford': 3652.095223880597,
 'mazda': 4010.7716643741405,
 'mercedes_benz': 8485.239571958942,
 'nissan': 4664.891034482758,
 'opel': 2876.716403542549,
 'peugeot': 3039.4682651622,
 'renault': 2395.4164467897976,
 'seat': 4296.492554410081,
 'skoda': 6334.91948051948,
 'smart': 3542.706586826347,
 'sonstige_autos': 10805.078393881453,
 'toyota': 5115.33388981636,
 'volkswagen': 5231.081983308787}

## Audi, BMW, and Benz are more expensive
## Ford and Open are less Expesnive
## Volkswagen is in between

In [68]:
avg_mileage = {}
for b in brand_types:
    brand_only = autos[autos['brand'] == b]
    avg_mileage[b] = int(brand_only['odometer_km'].mean())
    #avg_mileage[b] = brand_only['price'].mean()
    
avg_mileage

{'audi': 129287,
 'bmw': 132431,
 'citroen': 119580,
 'fiat': 116553,
 'ford': 124068,
 'mazda': 124745,
 'mercedes_benz': 130856,
 'nissan': 118572,
 'opel': 129223,
 'peugeot': 127136,
 'renault': 128183,
 'seat': 121563,
 'skoda': 110954,
 'smart': 99595,
 'sonstige_autos': 87466,
 'toyota': 115709,
 'volkswagen': 128724}

In [69]:
avg_price = {}
for b in brand_types:
    brand_only = autos[autos['brand'] == b]
    #avg_price[b] = brand_only['odometer_km'].mean()
    avg_price[b] = int(brand_only['price'].mean())
    
avg_price

{'audi': 9093,
 'bmw': 8102,
 'citroen': 3699,
 'fiat': 2711,
 'ford': 3652,
 'mazda': 4010,
 'mercedes_benz': 8485,
 'nissan': 4664,
 'opel': 2876,
 'peugeot': 3039,
 'renault': 2395,
 'seat': 4296,
 'skoda': 6334,
 'smart': 3542,
 'sonstige_autos': 10805,
 'toyota': 5115,
 'volkswagen': 5231}

In [70]:
brand_avg_mileage = pd.Series(avg_mileage) 
brand_avg_price = pd.Series(avg_price)
print(brand_avg_price)


audi               9093
bmw                8102
citroen            3699
fiat               2711
ford               3652
mazda              4010
mercedes_benz      8485
nissan             4664
opel               2876
peugeot            3039
renault            2395
seat               4296
skoda              6334
smart              3542
sonstige_autos    10805
toyota             5115
volkswagen         5231
dtype: int64


In [71]:
df = pd.DataFrame(brand_avg_mileage, columns = ['mean_mileage'])
df

mean_mileage
audi                  129287
bmw                   132431
citroen               119580
fiat                  116553
ford                  124068
mazda                 124745
mercedes_benz         130856
nissan                118572
opel                  129223
peugeot               127136
renault               128183
seat                  121563
skoda                 110954
smart                  99595
sonstige_autos         87466
toyota                115709
volkswagen            128724

In [72]:
df['mean_price'] = brand_avg_price

In [73]:
df

mean_mileage  mean_price
audi                  129287        9093
bmw                   132431        8102
citroen               119580        3699
fiat                  116553        2711
ford                  124068        3652
mazda                 124745        4010
mercedes_benz         130856        8485
nissan                118572        4664
opel                  129223        2876
peugeot               127136        3039
renault               128183        2395
seat                  121563        4296
skoda                 110954        6334
smart                  99595        3542
sonstige_autos         87466       10805
toyota                115709        5115
volkswagen            128724        5231

### Smart car has one of the lowest price and mean mileage
### Sonstige autos, Benz, Audi and BMW have the highest mean prices 